### ISSUE TITLE
Improve game name matching

### ISSUE THREAD

**Issue by RoorMakurosu** [gogcom/galaxy-integrations-python-api#95]
In many situations, game detection doesn't work very well.
For example, a game named "Dark Heart of Uukrul" doesn't match "The Dark Heart of Uukrul".. becoming an unknown game (and showing nothing in the game "dashboard").

Same as for situations when games have numbers... so "Master of Orion II" doesn't match "Master of Orion 2".

**Comment by RoorMakurosu** [comment]
Maybe applying the **levenshtein distance** would give excelent results to that.
See: [link](https://support.adeptia.com/hc/en-us/articles/207875583-Levenshtein-Algorithm-Fuzzy-Matching-)

**Comment by mbanczerowski** [comment]
Hi @RoorMakurosu 

"Unknown Game" should never appear I would consider this as our backend bug. If matching fails Galaxy should display a title the plugin reports. Can you provide zipped Galaxy logs if you still have this problem?

As far as I know matching algorithm cannot be too liberal, because it will produce many wrong results. Instead there is 2-step matching process:
- automatic match based on the title (quite strict as you noticed)
- manual curation of the rest, that includes creating non-existing game definitions and media as well as fixing false positives from automatic matches. Note that there are different games with the same title or very similar titles.

That is why it is usually takes some time until games ids sent from plugin are fully matched.

**Comment by burnhamup** [comment]
I'm working on an IndieGala plugin and triggered a few unknown games while building it. They all seem to be based on HTML entities. For example, I'm sending the following game from my plugin

{"game_id": "Another Lost Phone: Laura&#39;s Story", "game_title": "Another Lost Phone: Laura's Story", "dlcs": [], "license_info": {"license_type": "SinglePurchase"}}

I've attempted to strip the &#39; from the game_title, but had issues when using that same string as the game_id. I found the following in the  GalaxyClient.log

2020-04-24 21:32:32.160 [Warning][ (0)] [TID 24412][galaxy_client]: GamesDbPiecesPullerByRelease retrieved response code 404, will treat game 'indiegala_Another Lost Phone: Laura&#39;s Story' pieces as not set.

Does that matching happen on game title or game id? And is there guidance on how these should be formatted?

My plugin is here: https://github.com/burnhamup/galaxy-integration-indiegala

**Comment by RoorMakurosu** [comment]
Btw, a regular problem I experience...
It's that when I add a game manually, and it doesn't exist in the DB... I'm
forced to select a random game and rename it.

It would be cool to be able to add custom names.

Also, a much bigger issue...
I cannot add a game I already have installed, if I already own it in other
platform :'(

**Comment by Silphatos** [comment]
> Hi @RoorMakurosu
> 
> "Unknown Game" should never appear I would consider this as our backend bug. If matching fails Galaxy should display a title the plugin reports. Can you provide zipped Galaxy logs if you still have this problem?
> 
> As far as I know matching algorithm cannot be too liberal, because it will produce many wrong results. Instead there is 2-step matching process:
> 
> * automatic match based on the title (quite strict as you noticed)
> * manual curation of the rest, that includes creating non-existing game definitions and media as well as fixing false positives from automatic matches. Note that there are different games with the same title or very similar titles.
> 
> That is why it usually takes some time until games ids sent from plugin are fully matched.

Hello, is this issue gonna be fixed, almost all of the itch.io titles show up as "Unknown game" as these small indie games don't have IGDB pages usually.

**Comment by RoorMakurosu** [comment]
It would be good if in the GOG Galaxy API we could have a GAME TITLE and a different GAME "DB" TITLE.
Because sometimes the name in the IGDB pages are not a very good match.
For example, sometimes we got titles as "XXX Game Super Collector's Edition" when it is actually "XXX Game"...
Plex (an app for keeping video libraries) has a very nice solution for this, which allows you to manually select "movie" titles from a list of potential matches when the match made by the system was "not great".
On 4/11/2020 12:37:43, Silphatos <notifications@github.com> wrote:
Hi @RoorMakurosu [https://github.com/RoorMakurosu]
"Unknown Game" should never appear I would consider this as our backend bug. If matching fails Galaxy should display a title the plugin reports. Can you provide zipped Galaxy logs if you still have this problem?
As far as I know matching algorithm cannot be too liberal, because it will produce many wrong results. Instead there is 2-step matching process:
* automatic match based on the title (quite strict as you noticed)
* manual curation of the rest, that includes creating non-existing game definitions and media as well as fixing false positives from automatic matches. Note that there are different games with the same title or very similar titles.
That is why it usually takes some time until games ids sent from plugin are fully matched.
Hello, is this issue gonna be fixed, almost all of the itch.io titles show up as "Unknown game" as these small indie games don't have IGDB pages usually.

**Comment by mbanczerowski** [comment]
Hi, I'm passing the issue to attention of appropriate team.

**Comment by Silphatos** [comment]
Great! BTW those unknown games increase the client startup time somehow. Importing stuff gets stuck at 99% for a few minutes if the account has a lot of unknown games.

**Comment by RoorMakurosu** [comment]
In my case, it gets stuck at 99% for a couple of minutes everytime...
And even then, after reaching 100%, the application is "sluggish"  (super slow) for some more time; or well "crashes" and I'm forced to close and re-open it.
This happens like 50% of the time.
On 5/11/2020 08:32:28, Silphatos <notifications@github.com> wrote:
Great! BTW those unknown games increase the client startup time somehow. Importing stuff gets stuck at 99% for a few minutes if the account has a lot of unknown games.

**Comment by Silphatos** [comment]
Yes same here, I didn't have that problem until I added several unknown games with itch.io integration.

**Comment by RoorMakurosu** [comment]
I even make my own plugins for many platforms, and I know very well that they work "fast".
My library exceeds the 6000 titles so far. But only very few are unknown titles... even then, loading time have always been abyssal.

I partially blame the GOG Galaxy's log files. They write stuff to disc almost non-stop, and disc access is a big bottle-neck.
I would love to be able to disable those. Imagine, that a few lines are written into them (one per plugin, and a few for GOG galaxy basics) for every game in the library.
Logging is not very optimized, as it seems not to be buffered; and it seems to be written by each individual Python instance independently.

I haven't tried to see if that can be modified through the GOG Galaxy plugins API yet.


On 5/11/2020 09:28:31, Silphatos <notifications@github.com> wrote:
Yes same here, I didn't have that problem until I added several unknown games with itch.io integration.

**Comment by RoorMakurosu** [comment]
> I'm working on an IndieGala plugin and triggered a few unknown games while building it. They all seem to be based on HTML entities. For example, I'm sending the following game from my plugin
> 
> {"game_id": "Another Lost Phone: Laura's Story", "game_title": "Another Lost Phone: Laura's Story", "dlcs": [], "license_info": {"license_type": "SinglePurchase"}}
> 
> I've attempted to strip the ' from the game_title, but had issues when using that same string as the game_id. I found the following in the GalaxyClient.log
> 
> 2020-04-24 21:32:32.160 [Warning][ (0)] [TID 24412][galaxy_client]: GamesDbPiecesPullerByRelease retrieved response code 404, will treat game 'indiegala_Another Lost Phone: Laura's Story' pieces as not set.
> 
> Does that matching happen on game title or game id? And is there guidance on how these should be formatted?
> 
> My plugin is here: https://github.com/burnhamup/galaxy-integration-indiegala

I'm pretty sure that only the Game Title is used for matching it to the games DB.
I provide all my game Ids as hexadecimal values (a hash from the platform Id plus the full game title) and never caused any harm.

And the "GamesDbPiecesPullerByRelease" issue appears every time a game is not found in the DB... what I hate about it is that it seems to slow down the start up of GOG Galaxy a lot :'(  Would love that to be done a bit more "asynchronically".

**Comment by AndrewDWhite** [comment]
Here are the logs for the plugins with the greatest number of Unknown Game instances in my library. There are also some produced by the playstation plugin as well and I can reinstall that one to get the log info from it if that is helpful. One example is 828f89f54d4e56114dab8e2d8705a1f186440cd2 . The initial import log had aged off I can rerun that and attach a version with that information as well.

[GalaxyClient.log](https://github.com/gogcom/galaxy-integrations-python-api/files/7388964/GalaxyClient.log)
[plugin-test-97543122-7785-4444-2254-711233556699.log](https://github.com/gogcom/galaxy-integrations-python-api/files/7388966/plugin-test-97543122-7785-4444-2254-711233556699.log)
[plugin-itch-2df02142-4d8a-4a4b-9b6e-c3a0bc62f93b.log](https://github.com/gogcom/galaxy-integrations-python-api/files/7388967/plugin-itch-2df02142-4d8a-4a4b-9b6e-c3a0bc62f93b.log)

**Comment by AndrewDWhite** [comment]
Completely reinstalling everything, including os, and rewatching the initial process, as the import process is still running, here are the logs for the one called out example. 
[plugin-test-97543122-7785-4444-2254-711233556699.log](https://github.com/gogcom/galaxy-integrations-python-api/files/7389684/plugin-test-97543122-7785-4444-2254-711233556699.log)
Usually this process rolls over the logs and starts purging them for everything imported so I may not be able to get everything in one run.
